ministryofjustice/analytics-platform-control-panel-frontend
# E2E testing

## What

Implement end-to-end tests using [Codecept](https://codecept.io).

Since it's not possible to create users/superusers programmatically, a test user must be pre-created. This user must exist on GitHub, and be added to Auth0. [This user](https://github.com/analytics-test) has been created for the purpose.

The details for this user, including password and 2FA key need to be in your `.env` file, so get those before reviewing. (This user has been added to Auth0, but may on occasion need the Auth0 2FA QR code passing manually.)

Also, before you start, you will need the user to have been created in your local DB and have their email verified so that the email verification step does not occur during testing. To complete the 2FA step, use [authenticator](https://www.npmjs.com/package/authenticator) (this will need to be installed locally to the project anyway so that the tests can pass 2FA, so just `yarn` to install). Once this is installed use the following command to get a 2FA code:

`yarn authenticator --issuer GitHub --account TEST_USER_EMAIL --key TEST_USER_KEY`

replacing the relevant values with the values from `.env`. You will be given a 6-digit 2FA code in response. Once your user is authenticated and logged in, verify the email address in the control panel. You are then ready to run the tests.

## How to review

1. `yarn` to install `selenium-standalone` and `authenticator`
2. in a new tab `yarn selenium-standalone install` then `yarn selenium-standalone start`
3. run the non-superuser tests with `yarn codeceptjs run --steps --grep @nonsuperuser`
4. upgrade your user to superuser status using the API (I use Swagger)
5. run the superuser tests with `yarn codeceptjs run --steps --grep @superuser`

Each test suite should pass.

non-superuser:
  OK  | 2 passed   // 10s
✨  Done in 11.21s.

superuser:
  OK  | 61 passed   // 40s
✨  Done in 41.17s.
